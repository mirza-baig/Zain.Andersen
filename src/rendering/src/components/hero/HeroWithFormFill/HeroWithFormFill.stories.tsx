// Global
import { Story, Meta } from '@storybook/react';
// Local
import HeroWithFormFill, { HeroWithFormFillProps } from './HeroWithFormFill';
import defaultData from './HeroWithFormFill.mock-data';

export default {
  title: 'Hero/HeroWithFormFill',
  component: HeroWithFormFill,
} as Meta;
const Template: Story<HeroWithFormFillProps> = (props) => <HeroWithFormFill {...props} />;
export const Default = Template.bind({});
Default.args = defaultData;
