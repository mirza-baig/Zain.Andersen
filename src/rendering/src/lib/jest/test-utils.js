// Global
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ComponentPropsContext } from '@sitecore-jss/sitecore-jss-nextjs';
import { FastSitecoreContext } from 'lib/overrides/FastSitecoreContext';

// Lib
import { createSampleComponentFactory } from 'lib/mocks/mock-placeholder';

export const eeSitecoreContext = {
  route: {
    name: 'Foo',
    placeholders: {},
  },
  context: {
    pageEditing: true,
    site: {
      name: 'JssTestWeb',
    },
    language: 'en',
  },
};

export const renderComponent = (
  Component,
  {
    componentProps = {},
    staticProps = {},
    componentFactory = createSampleComponentFactory(),
    sitecoreContext = {
      route: {
        name: 'Foo',
        placeholders: {},
      },
    },
  }
) => {
  const mockLayoutData = {
    sitecore: sitecoreContext,
  };

  return render(
    <ComponentPropsContext value={staticProps}>
      <FastSitecoreContext componentFactory={componentFactory} layoutData={mockLayoutData}>
        <Component {...componentProps} />
      </FastSitecoreContext>
    </ComponentPropsContext>
  );
};

export const snapshot = (Component, mockData) => {
  const component = renderComponent(Component, mockData);
  expect(component.container.firstChild).toMatchSnapshot();
  return component;
};

export const hasDataComponent = async (component, name) => {
  await waitFor(() => {
    const nodes = component.baseElement.querySelectorAll(`[data-component="${name}"]`);
    return expect(nodes.length).toBe(1);
  });
};

export const mockDateAs = (epochMs) => {
  if (typeof epochMs !== 'number') {
    console.error(
      'mockDateAs only accepts epoch values in milliseconds to prevent timezones from factoring in to Date creation.'
    );
  }

  jest.spyOn(global.Date, 'now').mockImplementation(() => new Date(epochMs).valueOf());
};

export const testFieldError = async (Component, mockData, { label, value, error }) => {
  const component = renderComponent(Component, mockData);
  const field = component.getByLabelText(label);
  fireEvent.change(field, { target: { value } });
  field.blur();
  await waitFor(() => {
    const err = component.queryByText(error);
    return expect(err).not.toBe(null);
  });
};
