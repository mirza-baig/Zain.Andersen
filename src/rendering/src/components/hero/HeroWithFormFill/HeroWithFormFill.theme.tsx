// Lib
import { ThemeFile } from 'lib/context/ThemeContext';
import { FormAlignment } from './HeroWithFormFill';

const getFormWrapperClass = (alignment: FormAlignment): string => {
  switch (alignment) {
    case 'vertical':
      return 'ml:w-1/2 lg:min-w-[712px] ml:ml-auto';
    case 'horizontal':
      return 'w-full flex mx-auto mb-m ml:mt-l max-w-[1250px]';
    case 'horizontalSpillover':
      return 'ml:w-full flex ml:mx-auto max-w-[1250px] ml:-mt-[142px] ml:relative';
    default:
      return 'w-full';
  }
};

const getAlignmentClass = (alignment: FormAlignment): string => {
  switch (alignment) {
    case 'vertical':
      return 'ml:flex-row flex-col justify-center ml:p-l';
    case 'horizontal':
      return 'flex-col ml:p-m ml:min-h-[617px]';
    case 'horizontalSpillover':
      return 'flex-col ml:min-h-[617px] ml:p-m';
    default:
      return '';
  }
};

const getContentAlignmentClass = (alignment: FormAlignment): string => {
  switch (alignment) {
    case 'vertical':
      return '';
    case 'horizontal':
      return 'mx-auto ml:mt-auto max-w-[1250px] w-full';
    case 'horizontalSpillover':
      return 'ml:m-auto max-w-[1250px] w-full';
    default:
      return '';
  }
};

const getContentWidth = (alignment: FormAlignment): string => {
  switch (alignment) {
    case 'vertical':
      return 'ml:max-w-[490px]';
    case 'horizontal':
    case 'horizontalSpillover':
      return 'ml:max-w-[585px] mr-auto';
    default:
      return '';
  }
};

const getOverlayClass = (alignment: FormAlignment): string => {
  return alignment === 'horizontalSpillover' ? 'max-h-[617px]' : '';
};

export const HeroWithFormFillTheme = (alignment: FormAlignment): ThemeFile => {
  return {
    aw: {
      classes: {},
    },
    rba: {
      classes: {
        overlay: 'absolute z-[1] w-full h-full bg-[rgba(0,0,0,0.5)] left-0 top-0',
        contentWrapper: 'w-full h-full flex items-center ',
        contentContainer: 'flex flex-col z-[1] ml:ml-auto  max-ml:px-m  ',
        headline: 'my-xxxs text-sm-m ml:text-xxl ml:!text-white font-medium',
        bodyClass: 'my-xxxs text-body text-dark-gray ml:text-white ml:text-large-body',
        formWrapperClass: getFormWrapperClass(alignment),
        alignmentClass: getAlignmentClass(alignment),
        contentAlignmentClass: getContentAlignmentClass(alignment),
        contentwidth: getContentWidth(alignment),
        overlayClass: getOverlayClass(alignment),
      },
    },
  };
};
