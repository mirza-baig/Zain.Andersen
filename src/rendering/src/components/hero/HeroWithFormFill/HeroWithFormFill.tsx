import React from 'react';
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import { useTheme } from 'lib/context/ThemeContext';
import { withDatasourceCheck, Placeholder } from '@sitecore-jss/sitecore-jss-nextjs';
import classNames from 'classnames';
import { getEnum } from 'lib/utils';
import { getBreakpoint, useCurrentScreenType } from 'lib/utils/get-screen-type';
import { HeroWithFormFillTheme } from './HeroWithFormFill.theme';
import { Headline } from 'src/helpers/Headline';
import { MediaPrimary } from 'src/helpers/Media';
import RichTextWrapper from 'src/helpers/RichTextWrapper/RichTextWrapper';

export type FormAlignment = 'vertical' | 'horizontal' | 'horizontalSpillover';

export type HeroWithFormFillProps =
  Feature.EnterpriseWeb.RenewalByAndersen.Components.Hero.HeroWithFormFill.HeroWithFormFill;

const HeroWithFormFill: React.FC<HeroWithFormFillProps> = (props) => {
  const { rendering, fields } = props;
  const { currentScreenWidth } = useCurrentScreenType();
  const isDesktop = currentScreenWidth >= getBreakpoint('ml') ? true : false;
  const formAlignment = getEnum(fields?.formAlignment) as FormAlignment;
  const hasOverlay = getEnum(fields?.formStyle) === 'grayFormWithBlackOverlay';
  const { themeData } = useTheme(HeroWithFormFillTheme(formAlignment));
  if (!fields) {
    return null;
  }

  const primaryImage = fields?.primaryImage?.value;
  const imageStyles =
    primaryImage && isDesktop
      ? {
          backgroundImage: `url(${primaryImage.src})`,
          backgroundSize: 'cover',
          backgroundPosition: 'center',
        }
      : {};

  const formBGClass = hasOverlay
    ? 'bg-light-gray'
    : 'bg-black [&_div]:!text-white [&_input]:!text-black theme-black';

  const renderForm = rendering?.placeholders?.form?.length;

  return (
    <div data-component="hero/herowithformfill" className="relative h-full w-full" {...props}>
      {hasOverlay && isDesktop && (
        <div
          className={classNames(themeData.classes.overlay, themeData.classes.overlayClass)}
        ></div>
      )}

      <div
        className={classNames(themeData.classes.contentWrapper, themeData.classes.alignmentClass)}
        style={imageStyles}
      >
        {/* content */}
        <div
          className={classNames(
            themeData.classes.contentContainer,
            themeData.classes.contentAlignmentClass
          )}
        >
          <Headline
            classes={classNames(themeData.classes.headline, themeData.classes.contentwidth)}
            {...props}
          />
          <RichTextWrapper
            classes={classNames(
              themeData.classes.bodyClass,
              themeData.classes.contentwidth,
              'ml:[&>div]:!text-white'
            )}
            field={fields?.body}
          />
        </div>

        {/* primaryImage in mobile */}
        <MediaPrimary
          imageLayout="fill"
          ratio="hero"
          additionalDesktopClasses="ml:!hidden"
          additionalMobileClasses="!block md:!block min-h-[280px] mt-l ml:!hidden"
          priority
          {...props}
        />
        {/* vertical or horizontal Form */}
        {renderForm && formAlignment !== 'horizontalSpillover' && (
          <div className={classNames('z-[1] p-s', themeData.classes.formWrapperClass, formBGClass)}>
            <Placeholder
              name="form"
              rendering={rendering}
              isHorizontalForm={formAlignment === 'horizontal'}
            />
          </div>
        )}
      </div>
      {/* horizontalSpillover Form */}
      {renderForm && formAlignment === 'horizontalSpillover' && (
        <div className={classNames('z-[1] p-s', themeData.classes.formWrapperClass, formBGClass)}>
          <Placeholder name="form" rendering={rendering} isHorizontalForm={true} />
        </div>
      )}
    </div>
  );
};

export default withDatasourceCheck()<HeroWithFormFillProps>(HeroWithFormFill);
