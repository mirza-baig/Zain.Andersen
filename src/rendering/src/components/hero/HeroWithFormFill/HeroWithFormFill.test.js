// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
// Local
import HeroWithFormFill from './HeroWithFormFill';
import defaultData from './HeroWithFormFill.mock-data';

it('renders correctly', async () => {
  const component = snapshot(HeroWithFormFill, { componentProps: defaultData });
  await hasDataComponent(component, 'hero/herowithformfill');
});
