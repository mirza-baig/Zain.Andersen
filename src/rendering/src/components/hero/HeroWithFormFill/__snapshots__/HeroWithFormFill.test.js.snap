// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="relative h-full w-full"
  data-component="hero/herowithformfill"
  fields="[object Object]"
  rendering="[object Object]"
>
  <div
    class=""
  />
  <div
    class=""
    style="background-image: url(https://edge.sitecorecloud.io/andersencorporation-m9n8s20s/media/Project/AndersenCorporation/RenewalByAndersen/RenewalByAndersen/Images/Hero-Media-Background/why-andersen-hero-image1400x617.jpg?h=617&iar=0&w=1440&rev=6e4d257fc1e643c882e2795926accb7e); background-size: cover; background-position: center;"
  >
    <div
      class=""
    >
      <div
        class="items-top font-sans [&_a:hover]:underline text-theme-text text-sm-m md:text-m font-bold mb-s"
      >
        <h2>
          This is a headline
        </h2>
      </div>
      <div
        class="body-copy ml:[&>div]:!text-white"
      >
        <div>
          Lorem ipsum dolor sit amet, consectetur adipisicing elit. Delectus earum perspiciatis quibusdam porro voluptatum cum laboriosam laudantium magni modi! Ipsam eos labore voluptatum illum reiciendis amet nisi, dolorem soluta quisquam.
        </div>
      </div>
    </div>
    <div
      class="relative mx-auto hidden md:block aspect-hero   ml:!hidden"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="why-andersen-hero-image1400x617"
          data-nimg="fill"
          decoding="async"
          sizes="100vw"
          src="/_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=3840&q=75"
          srcset="/_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=640&q=75 640w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=750&q=75 750w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=828&q=75 828w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=1080&q=75 1080w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=1200&q=75 1200w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=1920&q=75 1920w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=2048&q=75 2048w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=3840&q=75 3840w"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
        />
      </span>
    </div>
    <div
      class="relative w-full md:hidden !block md:!block min-h-[280px] mt-l ml:!hidden"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="why-andersen-hero-image1400x617"
          data-nimg="fill"
          decoding="async"
          sizes="100vw"
          src="/_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=3840&q=75"
          srcset="/_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=640&q=75 640w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=750&q=75 750w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=828&q=75 828w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=1080&q=75 1080w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=1200&q=75 1200w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=1920&q=75 1920w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=2048&q=75 2048w, /_next/image/?url=https%3A%2F%2Fedge.sitecorecloud.io%2Fandersencorporation-m9n8s20s%2Fmedia%2FProject%2FAndersenCorporation%2FRenewalByAndersen%2FRenewalByAndersen%2FImages%2FHero-Media-Background%2Fwhy-andersen-hero-image1400x617.jpg%3Fh%3D617%26iar%3D0%26w%3D1440%26rev%3D6e4d257fc1e643c882e2795926accb7e&w=3840&q=75 3840w"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover; object-position: center;"
        />
      </span>
    </div>
    <div
      class="z-[1] p-s bg-light-gray"
    >
      <div
        style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
      >
        <h2>
          Form
        </h2>
        <p>
          JSS component is missing React implementation. See the developer console for more information.
        </p>
      </div>
    </div>
  </div>
</div>
`;
